### ISSUE TITLE
Repair requires large *contiguous* allocation when there are large differences

### ISSUE THREAD

**Issue by nyh** [scylladb/scylla#2412]
Scylla version (or git commit hash): 1.7

Since commit 66e3b73b9c7aee32a9dcbd1f3434610498155f9a, we break up the repaired range into small ranges - around 100 partitions. But a consequence of this is that if there are very large differences (e.g., in Roy's test we deliberately remove entire sstables from one of the nodes), we will get a huge number of differing ranges. We do have a limit on this number of ranges we may collect - size_t sub_ranges_max = 10 * 1024 * 1024, but this is a large number (10 million) and I think we actually can get this multiplied by number of vnodes for the total number of subranges which will need to be streamed.

Making a list of 10 or 100 million objects doesn't sound prohibitive, but it is when we require this list to be consecutive, i.e., std::vector... We simply may not have that much consecutive free memory, and fail with bad_alloc when trying to append items to this list.

"ranges" in repair_cf_range() could be a problem, but it appears there is a much bigger problem: the stream_plan, to which we add the entire huge list of all the ranges and nodes which differed. When there are very big differences (as in the test Roy tried), this list can contain millions of items. Looking at streaming/stream_session.hh, the requests are stored in a std::vector<stream_request> _requests. This means that all these millions of range requests are listed in contiguous memory - and we can fail such allocation, and apparently we do (in Roy's test we saw bad_alloc in the phase of streaming which adds ranges to be synced).

I think an easy solution to this issue could be to use seastar::chunked_fifo instead of std::vector for stream_session::_requests.

A more thorough solution could be for repair NOT to try to build one stream plan with all the millions of differing ranges...  Start the stream after collecting a few ranges, there is no reason to collect a million of them... This is how the code was organized earlier, by the way :-) But we should probably not bother with this solution at this point because we want to rewrite the repair. But let's remember this point in the new repair implementation, and never go over all the data to make one giant list.

**Comment by avikivity** [comment]
Let's have a quick fix ASAP, then a more involved fix, this can be hitting users.

**Comment by nyh** [comment]
I'll produce now a patch to use seastar::chunked_fifo and ask @asias to please review.

**Comment by nyh** [comment]
By the way, an alternative possible fix: sp_in.request_ranges() et al. already allow you to give a vector of ranges, not just one range. So we could collect this vector and make one call to sp_in.request_ranges() instead of making a million calls to sp_in.request_ranges(). 
Should be easier to implement (especially for me, who's not familiar with the streaming code internals).

**Comment by roydahan** [comment]
@nyh if you can provide me a patch on top of 1.7.0, I'll be able to test it much quicker.

**Comment by asias** [comment]
@nyh we can indeed merge the same ranges under the same keyspace and cf into a single stream_request which can save memory to store the keyspace and cf.

**Comment by nyh** [comment]
@gleb-cloudius noted that while converting vectors to list can help this issue, there will remain another contiguous allocation problem: The uncompressed RPC layer currently keeps an entire message in memory, but not contiguously, so it's fine. However, the **compressed** (and encrypted?) RPC variants put the entire message in contiguous memory - and this message can be very large.

I (@nyh) noticed that we have yet another contiguous vector - in repair_cf_range() we have `vector<::dht::token_range> ranges`, which may contain up to 10 million items (in the worst case, although this will require billions of partitions in the table).

**Comment by nyh** [comment]
Note that about repair_cf_range's "ranges" in my last comment - while I was right in the worst case, because repair_cf_range is called separately for each vnode, it actually allocates just 1/256 (vnodes=256) of the list each time. So for example if we have a huge list spanning 3 gigabytes of memory just for the list of 100-partition ranges, we will only allocate 10 MB contiguous memory for each vnode, and that is more manageable than trying to allocate the whole 3 gigabytes at once.
The bugs we discussed above were more serious because we kept a list for the entire node (not each vnode separately) consecutively.
